pgmpy/inference/mplp.py

```diff
@@ -0,0 +1,172 @@
+
+    class Cluster:
+
+        """
+        Inner class for representing a cluster.
+        A cluster is a subset of variables.
+
+        Parameters
+        ----------
+        set of variables: tuple
+                          This is the set of variables that form the cluster.
+
+        intersection_set_variables: set containing frozensets.
+                                    collection of intersection of all pairs of cluster variables.
+                                    For eg: {{C1 ∩ C2}, {C2 ∩ C3}, {C3 ∩ C1}} for clusters C1, C2 & C3.
+
+        cluster_potential: Factor
+                           Each cluster has a initial probability distribution provided beforehand.
+        """
+        def __init__(self, set_of_variables_for_cluster_c, intersection_set_variables, cluster_potential):
+            """
+            Initialization of the current cluster
+            """
+
+            # The variables with which the cluster is made of.
+            self.cluster_variables = set_of_variables_for_cluster_c
+
+            # The cluster potentials must be specified before only.
+            self.cluster_potential = copy.deepcopy(cluster_potential)
+
+            # Generate intersection sets for this cluster; S(c)
+            self.intersection_sets_for_cluster_c = [i.intersection(self.cluster_variables)
+                                                    for i in intersection_set_variables
+                                                    if i.intersection(self.cluster_variables)]
+
+            # Initialize messages from this cluster to its respective intersection sets
+            # \lambda_{c \rightarrow \s} = 1/|S(c)| * max_{x_{c-s}}{\theta_{ij}}
+            self.message_from_cluster = {}
+            for i in self.intersection_sets_for_cluster_c:
+                other_variables = list(set(set_of_variables_for_cluster_c)-i)
+                phi = copy.deepcopy(cluster_potential)
+
+                # phi = max_{x_{c-s}}{\theta_{ij}}
+                phi.maximize(other_variables)
+                self.message_from_cluster[i] = (1 / len(self.intersection_sets_for_cluster_c)) * phi
+
+    def _update_message(self, sending_cluster):
+
+        """
+        This is the message-update method.
+
+        Parameters
+        ----------
+        sending_cluster: The resulting messages are \lambda_{c \rightarrow s} from the given
+            cluster 'c' to all of its intersection_sets 's'.
+            Here 's' are the elements of intersection_sets_for_cluster_c.
+
+        Reference
+        ---------
+        Fixing Max-Product: Convergent Message-Passing Algorithms for MAP LP Relaxations
+        by Amir Globerson and Tommi Jaakkola.
+        Section 6, Page: 5; Beyond pairwise potentials: Generalized MPLP
+        """
+        updated_results = []
+        # The new updates will take place for the intersection_sets of this cluster.
+        for i in sending_cluster.intersection_sets_for_cluster_c:
```

Comment:
Please use appropriate variables instead of ```i```. In future don't use any variable using i, j, k, l; use proper variable names.